shell/platform/linux/fl_basic_message_channel.cc

```diff
@@ -0,0 +1,217 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_basic_message_channel.h"
+
+#include <gmodule.h>
+
+typedef struct {
+  // Messenger to communicate on
+  FlBinaryMessenger* messenger;
+
+  // Channel name
+  gchar* name;
+
+  // Codec to en/decode messages
+  FlMessageCodec* codec;
+
+  // Function called when a message is received
+  FlBasicMessageChannelMessageHandler message_handler;
+  gpointer message_handler_data;
+} FlBasicMessageChannelPrivate;
+
+// Added here to stop the compiler from optimising this function away
+G_MODULE_EXPORT GType fl_basic_message_channel_get_type();
+
+enum { PROP_MESSENGER = 1, PROP_NAME, PROP_CODEC, PROP_LAST };
+
+G_DEFINE_TYPE_WITH_PRIVATE(FlBasicMessageChannel,
+                           fl_basic_message_channel,
+                           G_TYPE_OBJECT)
+
+// Called when a binary message is received on this channel
+static void message_cb(FlBinaryMessenger* messenger,
+                       const gchar* channel,
+                       GBytes* message,
+                       FlBinaryMessengerResponseHandle* response_handle,
+                       gpointer user_data) {
+  FlBasicMessageChannel* self = FL_BASIC_MESSAGE_CHANNEL(user_data);
+  FlBasicMessageChannelPrivate* priv =
+      static_cast<FlBasicMessageChannelPrivate*>(
+          fl_basic_message_channel_get_instance_private(self));
+
+  if (priv->message_handler == nullptr) {
+    fl_binary_messenger_send_response(messenger, response_handle, nullptr,
+                                      nullptr);
+    return;
+  }
+
+  g_autoptr(GError) error = nullptr;
+  g_autoptr(FlValue) message_value =
+      fl_message_codec_decode_message(priv->codec, message, &error);
+  if (message_value == nullptr)
+    g_warning("Failed to decode message: %s", error->message);
+  else
+    priv->message_handler(self, message_value, priv->message_handler_data);
+
+  fl_binary_messenger_send_response(messenger, response_handle, nullptr,
```

Comment:
This should be in the `if` branch, unless I'm missing something. `message_handler` should be sending the response if it was called.